### PR TITLE
Make the category deletion process clearer for the end user

### PR DESCRIPTION
**Pull request by DaazKu**
Fixes https://github.com/vanilla/vanilla/issues/5037

You now have to choose between moving the content or deleting the whole category and all its content.
The proceed button is disabled until you choose an appropriate action or until you checked all required field.
A better warning plus a confirmation checkbox about what will happen when you delete the category has been added.

### CODE REVIEW

File: applications/vanilla/controllers/class.vanillasettingscontroller.php [hunk 1]
```diff
@@ -488,14 +488,38 @@ public function deleteCategory($CategoryID = false) {
...
                     $this->Form->addError('You cannot remove the only remaining category that allows discussions');
+                } else {
+                    $newCategoryID = 0;
+                    $contentAction = $this->Form->getFormValue('ContentAction', false);
+
+                    switch($contentAction) {
+                        case 'move':
+                            $newCategoryID = $this->Form->getFormValue('ReplacementCategoryID');
+                            if (!$newCategoryID) {
+                                $this->Form->addError('A replacement category should have been selected!');
```
Comment (linc) [inline applications/vanilla/controllers/class.vanillasettingscontroller.php]: Replacement category is required.

File: applications/vanilla/controllers/class.vanillasettingscontroller.php [hunk 2]
```diff
@@ -488,14 +488,38 @@ public function deleteCategory($CategoryID = false) {
...
+                            }
+                            break;
+                        case 'delete':
+                            if (!$this->Form->getFormValue('ConfirmDelete', false)) {
+                                $this->Form->addError('The confirmation box was not checked!');
```
Comment (linc) [inline applications/vanilla/controllers/class.vanillasettingscontroller.php]: You must confirm the deletion.

File: applications/vanilla/views/vanillasettings/deletecategory.php [hunk 3]
```diff
@@ -41,9 +48,40 @@ class="text-danger"><?php echo t('<strong>Heads Up!</strong> Moving discussions
                 ?>
                 </div>
             </li>
+            <li id="DeleteCategory">
+            <?php if ($this->data('DiscussionsCount') || $subcategories) { ?>
+                <div class="alert alert-danger padded">
+                <?php if ($this->data('DiscussionsCount')) { ?>
+                    <p>
+                        <?php printf(t('<strong>%s</strong> discussion(s) will be deleted. There will be no way to restore them. They will not go in the changelog!'), $this->data('DiscussionsCount')); ?>
```
Comment (linc) [inline applications/vanilla/views/vanillasettings/deletecategory.php]: Use `plural()`.

Comment (linc) [inline applications/vanilla/views/vanillasettings/deletecategory.php]: There is no undo and they will not be logged.

File: applications/vanilla/views/vanillasettings/deletecategory.php [hunk 4]
```diff
@@ -41,9 +48,40 @@ class="text-danger"><?php echo t('<strong>Heads Up!</strong> Moving discussions
...
+                    </p>
+                <?php
+                }
+                if ($subcategories) {
+                ?>
+                    <p>
+                        <?php printf(t('<strong>%s</strong> sub-category(ies) will be deleted!'), count($subcategories)); ?>
```
Comment (linc) [inline applications/vanilla/views/vanillasettings/deletecategory.php]: Use `plural()`, and "child categories" is the term we use officially. No bang.

File: applications/vanilla/views/vanillasettings/deletecategory.php [hunk 5]
```diff
@@ -13,24 +18,26 @@
...
     } else { ?>
-        <div class="alert alert-danger padded"><?php echo t('All content in this category will be permanently deleted.'); ?></div>
         <ul>
             <li class="form-group">
                 <div class="input-wrap">
-                <?php echo $this->Form->checkBox('MoveContent', "Move content in this category to a replacement category.", ['value' => '1']); ?>
+                <?php echo $form->radio('ContentAction', 'Move content from this category to a replacement category.', ['value' => 'move']); ?>
+                </div>
+                <div class="input-wrap">
+                <?php echo $form->radio('ContentAction', 'Delete all the content in this category.', ['value' => 'delete']); ?>
```
Comment (linc) [inline applications/vanilla/views/vanillasettings/deletecategory.php]: Permanently delete all content in this category.

File: applications/vanilla/views/vanillasettings/deletecategory.php [hunk 6]
```diff
@@ -41,9 +48,40 @@ class="text-danger"><?php echo t('<strong>Heads Up!</strong> Moving discussions
...
+                    <p>
+                        <?php printf(t('<strong>%s</strong> discussion(s) will be deleted. There will be no way to restore them. They will not go in the changelog!'), $this->data('DiscussionsCount')); ?>
+                    </p>
+                <?php
+                }
+                if ($subcategories) {
+                ?>
+                    <p>
+                        <?php printf(t('<strong>%s</strong> sub-category(ies) will be deleted!'), count($subcategories)); ?>
+                        <a tabindex="0" role="button" class="js-category-list-toggle" data-show="Show the category list" data-hide="Hide the category list">Show the category list</a>.
```
Comment (linc) [inline applications/vanilla/views/vanillasettings/deletecategory.php]: Strike "the" from both messages

File: applications/vanilla/views/vanillasettings/deletecategory.php [hunk 7]
```diff
@@ -41,9 +48,40 @@ class="text-danger"><?php echo t('<strong>Heads Up!</strong> Moving discussions
...
+                                <li><?php echo $subcategory['Name'] ?></li>
+                            <?php endforeach; ?>
+                        </ul>
+                    </div>
+                <?php } ?>
+                </div>
+            <?php } ?>
+                <div class="form-group">
+                    <div class="input-wrap">
+                        <?php echo $form->checkBox('ConfirmDelete', 'I understand and want to delete the category and all its content.'); ?>
```
Comment (linc) [inline applications/vanilla/views/vanillasettings/deletecategory.php]: Yes, permanently delete it all.

**Comment by linc** [comment]
screen shot 2017-01-26 at 4 08 03 pm

This doesn't look right. Can we get an assist for what needs correcting here @beckyvb ?

**Comment by beckyvb** [comment]
I don't think we have any existing patterns that match it. It needs to go through @danmaitland First guess is that everything past and including "4 child categories will be deleted... " should be outside the alert div.

**Comment by linc** [comment]
I'm OK with not enumerating the child categories if it's gonna be a whole ordeal.

**Comment by beckyvb** [comment]
Feedback from Dan: 
* Remove show/hide categories toggle and list them all by default. 
* Keep the category list in the red alert. 
* Make sure there are bullet points on the li items. (The list-style is removed on forms, modal necessarily wrap everything in a form to get the submit buttons into the footer. Use the .list-un-reset CSS class on the list.)

**Comment by DaazKu** [comment]
That is not an acceptable solution. The list can potentially have 100 items....

**Comment by beckyvb** [comment]
It's better that they review all the categories that will be deleted. If there's 100, then they'll have to scroll a little. And during the time they spend scrolling, they'll fully comprehend what they'll be deleting.

**Comment by DaazKu** [comment]
That's actually a good point! Making the changes now!

**Pull request merged**